### PR TITLE
MULE-8654: Setting muleContext to internal transformers.

### PR DESCRIPTION
**Pull request by afelisatti**

### CODE REVIEW

File: core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java [hunk 1]
```diff
@@ -90,6 +86,7 @@ public MuleEvent answer(InvocationOnMock invocationOnMock) throws Throwable
         muleLockFactory.initialise();
         when(mockMuleContext.getLockFactory()).thenReturn(muleLockFactory);
         when(mockMuleContext.getRegistry().get(MuleProperties.OBJECT_STORE_MANAGER)).thenReturn(mockObjectStoreManager);
+        when(mockMuleContext.getConfiguration().getDefaultEncoding()).thenReturn(UTF_8);
```
Comment (marianogonzalez) [inline core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java]: how is this change related to the issue?

File: core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java [hunk 2]
```diff
@@ -102,6 +103,8 @@ public void initialise() throws InitialisationException
         idrId = String.format("%s-%s-%s",appName,flowName,"idr");
         lockFactory = muleContext.getLockFactory();
         store = createStore();
+        LifecycleUtils.initialiseIfNeeded(objectToByteArray, muleContext);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java]: static imports

File: transports/ftp/src/test/java/org/mule/transport/ftp/reliability/FtpRedeliveryPolicyTestCase.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public static Collection<Object[]> parameters()
+    {
+        return Arrays.asList(new Object[][] {
+                {ConfigVariant.FLOW, "reliability/ftp-redelivery-policy-config.xml"}
+        });
+    }
+    @Test
+    public void testRedeliveryPolicyDLQConsumesMessage() throws Exception
+    {
+        ExceptionListener exceptionListener = new ExceptionListener(muleContext).setNumberOfExecutionsRequired(3);
```
Comment (marianogonzalez) [inline transports/ftp/src/test/java/org/mule/transport/ftp/reliability/FtpRedeliveryPolicyTestCase.java]: aawwweeessooomeeeeee.... didn't know we had this class :)

Comment (marianogonzalez) [inline transports/ftp/src/test/java/org/mule/transport/ftp/reliability/FtpRedeliveryPolicyTestCase.java]: I'd be great if instead of hardcoding a 3 here and a 2 in the flow, you could define a MAX_REDELIVERY_ATTEMPTS constant and expose it as a system property. So, in here you would use MAX_REDELIVERY_ATTEMPS +1 and in the flow ${MAX_REDELIVERY_ATTEMPS}. In this way, you don't spread magic numbers across both the test class and the configuration.

Something like this is implemented in AbstractSftpRedeliveryTestCase

File: core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java (hunk 1 above)
Comment (afelisatti) [inline core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java]: When the transform method is executed, the encoding is determined. Since muleContext was null before, it was determined by the mule.encoding system property instead (see SystemUtils getDefaultEncoding method).

File: transports/ftp/src/test/java/org/mule/transport/ftp/reliability/FtpRedeliveryPolicyTestCase.java (hunk 3 above)
Comment (afelisatti) [inline transports/ftp/src/test/java/org/mule/transport/ftp/reliability/FtpRedeliveryPolicyTestCase.java]: Nice. Thanks.

File: core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java (hunk 1 above)
Comment (marianogonzalez) [inline core/src/test/java/org/mule/processor/IdempotentRedeliveryPolicyTestCase.java]: perfect. I didn't see that one coming. Thanks
